### PR TITLE
Revamp Documentation

### LINKED ISSUE #7: Suggestion: screenshots / gifs in the readme
For a UI lib, you really need screenshots and gifs in the README. Show me what it looks like! That will make me interested.

Just my opinion.

### PR DESCRIPTION
**Pull request by vuki656**
- [x] Popup
- [x] Select
- [x] Input
- [x] Split

---

I've cleaned up the docs for the popup. I left a few TODOs where I didn't know what to put so please let me know what goes there.

Before I do the remaining components, I'd like your input. 

I'm open to suggestions in regards to formatting and arranging everything so please provide feedback if you have any.

[HERE](https://github.com/vuki656/nui.nvim) is how the home page `.md` looks like and [HERE](https://github.com/vuki656/nui.nvim/blob/main/doc/components/popup.md) is the full popup API

Closes #7

### CODE REVIEW

File: README.md
Comment (MunifTanjim) [inline README.md:28]: Let's use the same indentation as the codebase (2 spaces).

File: doc/components/popup.md
Comment (MunifTanjim) [inline doc/components/popup.md:342]: Describing these here seems unnecessary, since `buf_options` can take other buffer optinos too (not just these two).

File: doc/components/popup.md
Comment (MunifTanjim) [inline doc/components/popup.md:365]: Describing these here seems unnecessary, since `win_options` can take other window optinos too (not just these two).

File: doc/components/popup.md
Comment (MunifTanjim) [inline doc/components/popup.md:53]: 🤔 Umm... This format seems a bit repetitive. How do you feel about changing it a little bit like this:


```md
**Type**: `table`

Contains all border related options
```

File: doc/components/popup.md
Comment (MunifTanjim) [inline doc/components/popup.md:73]: - Can we call these "Examples" instead of "Usage"?
- Nesting markdown headers so deep seems unnecessary? Maybe we can just do `**Examples**` instead?

File: doc/components/popup.md
Comment (MunifTanjim) [inline doc/components/popup.md:76]: Can we change these into `border.highlight` and so on?

File: doc/components/popup.md
Comment (MunifTanjim) [inline doc/components/popup.md:195]: These headers are way too deep (6 levels) 😅 Since each component will have separate documentation file, I think we can keep markdown headers within 3 levels.

File: doc/components/popup.md
Comment (MunifTanjim) [inline doc/components/popup.md:6]: Let's call these "Options" (like it's called in the codebase) instead of "Properties".

File: doc/components/popup.md
Comment (MunifTanjim) [inline doc/components/popup.md:1]: Does vim `:help` works with markdown files? 🤔 (probably not as far as I know)...

If not, then we should better put these files close to the source code instead, for example: `lua/nui/popup/README.md`.

For components without separate directory, you can move them (e.g. move `lua/nui/input.lua` into `lua/nui/input/init.lua`) and add the `README.md` file in there.

**Review by MunifTanjim — CHANGES_REQUESTED** [review]
Hey @vuki656, thank you so much for spending the time to do this! I really appreciate it 😃 

I hope you don't get annoyed, since I'm adding a lots of comments here 😅 Feel free to share if you don't agree with some of them...

File: README.md
Comment (vuki656) [reply README.md:28]: Makes sense, will fix.

File: doc/components/popup.md
Comment (vuki656) [reply doc/components/popup.md:342]: Ohh, I didn't know it accepted other options. Could we point a user to a list of those options if they are in the help somewhere?

File: doc/components/popup.md
Comment (vuki656) [reply doc/components/popup.md:365]: Same answer as above, would be nice to point the user to a list of all of them if possible.

File: doc/components/popup.md
Comment (vuki656) [reply doc/components/popup.md:53]: Makes sense, I was trying a lot of iterations and this one seemed kinda okay but I get your point.

File: doc/components/popup.md
Comment (vuki656) [reply doc/components/popup.md:73]: Sure thing.

The reason I was using markdown headers a lot is I have markdownlint enabled and as far as I know it "enforces" some standard markdown guidelines and those are don't bold something that should represent a title.

Thats the only reason pretty much and it is kinda hard to read so I'm down to switch back to bolded.

File: doc/components/popup.md
Comment (vuki656) [reply doc/components/popup.md:76]: Yup, that makes more sense than nesting things. Cool tip.

File: doc/components/popup.md
Comment (vuki656) [reply doc/components/popup.md:195]: Will address these by converting some of them to bolded text.

File: doc/components/popup.md
Comment (vuki656) [reply doc/components/popup.md:1]: I don't think it works with markdown yup.

Hmm, I think it would be easier for the user to just pop into the `doc` folder and find all the docs there. 

This way you will have one readme per component and the user will have to go back and forth if they want to read it. True that there will be an option to visit all of them from the main readme but still I think them being all in one folder Is simpler. 

Plus it doesn't clutter up the source code with unnecessary files.

But if you prefer readmes in the directory directly I'll move them no problem.

**Comment by vuki656** [comment]
> Hey @vuki656, thank you so much for spending the time to do this! I really appreciate it 
> 
> I hope you don't get annoyed, since I'm adding a lots of comments here  Feel free to share if you don't agree with some of them...

Not at all, that's why I only did the popup so we can discuss things in a draft. 

Will probably push the changes within 2-3 days as I'm traveling.

File: doc/components/popup.md
Comment (MunifTanjim) [reply doc/components/popup.md:342]: Do `:help options` and search for `local to buffer`

File: doc/components/popup.md
Comment (MunifTanjim) [reply doc/components/popup.md:365]: Do `:help options` and search for `local to window`

File: doc/components/popup.md
Comment (MunifTanjim) [reply doc/components/popup.md:1]: Well, keeping it with the source and naming them "README.md" has benefits actually.

From the user's perspective - they can option `lua/nui/popup` folder on GitHub and GitHub will render the `README.md` file automatically (without needing to open the markdown file separately). Without the name `README.md` it won't happen.

From the maintainer's perspective - it's easier to edit the markdown file in the same folder as the source code, instead of editing the markdown file from a different folder altogether.

File: doc/components/popup.md
Comment (MunifTanjim) [inline doc/components/popup.md:407]: `opts` is a table with these keys and boolean values:

https://github.com/MunifTanjim/nui.nvim/blob/abd1a4a23d0bbc0d0d4bc8faaa240330b44a4d5c/lua/nui/popup/init.lua#L265

File: doc/components/popup.md
Comment (MunifTanjim) [inline doc/components/popup.md:408]: `force` is a boolean: `true`/`false`

https://github.com/MunifTanjim/nui.nvim/blob/abd1a4a23d0bbc0d0d4bc8faaa240330b44a4d5c/lua/nui/popup/init.lua#L266

File: doc/components/popup.md
Comment (MunifTanjim) [inline doc/components/popup.md:428]: This can be a list of event names or just a single event name, or even comma separated event names. So all this works:

```lua
{ event.BufLeave, event.BufDelete }
-- or
{ event.BufLeave, "BufDelete" }
-- or
event.BufLeave
-- or
"BufLeave"
-- or
"BufLeave,BufDelete"
```

File: doc/components/popup.md
Comment (MunifTanjim) [inline doc/components/popup.md:430]: `options` is optional. It can be a table with the keys `once`, `nested` and boolean values.

https://github.com/MunifTanjim/nui.nvim/blob/abd1a4a23d0bbc0d0d4bc8faaa240330b44a4d5c/lua/nui/utils/autocmd.lua#L279

File: doc/components/popup.md
Comment (vuki656) [reply doc/components/popup.md:342]: What could we write here so users can easily figure it out?

I went in options help and searched for local to buffer and it wasn't as straightforward as it could be.

Comment (MunifTanjim) [reply doc/components/popup.md:337]: Well, I don't think there are any separate list for these options... they're all mixed together in `:help options`. So nothing much we can do here.

You can run `:help options | /local to buffer` and the search will be initiated for you. Then you'll just have to press `n` to go to next results.

**Comment by vuki656** [comment]
Hey @MunifTanjim I think I addressed all the points, can you take a look again?

File: lua/nui/popup/popup.md
Comment (MunifTanjim) [inline lua/nui/popup/popup.md:332]: Just call it `Methods` (like `Options`)?

File: lua/nui/popup/popup.md
Comment (MunifTanjim) [inline lua/nui/popup/popup.md:365]: These can be just bold texts instead of header. These headers look like this in Table of Content:

image

File: lua/nui/popup/popup.md
Comment (MunifTanjim) [inline lua/nui/popup/popup.md:1]: Rename the file to `README.md`. Otherwise GitHub won't render the markdown file when you're in `lua/nui/popup` folder:

image

Renaming it to `README.md` will render the markdown file below the file list... just like you see it on the homepage:

image

**Review by MunifTanjim — COMMENTED** [review]
Just need some minor adjustments, otherwise looks good 👍🏼

File: lua/nui/popup/popup.md
Comment (vuki656) [reply lua/nui/popup/popup.md]: Do you mean for all of them or just his particular one? 

I find bullet points a lot easier to read as its just bite-sized statements instead of bigger paragraphs.

But again, up to you.

Comment (MunifTanjim) [reply lua/nui/popup/popup.md]: > Do you mean for all of them or just his particular one?

all of them

> I find bullet points a lot easier to read as its just bite-sized statements instead of bigger paragraphs.

You got a point there. But these information are not really suitable for lists... seeing them in list seems kinda forced. 🤔 Also the paragraphs doesn't need to be big. You can have bite-sized paragraphs as well. 😃

**Comment by MunifTanjim** [comment]
Everything looking good for `popup` 🚀 @vuki656

**Comment by vuki656** [comment]
> Everything looking good for `popup`  @vuki656

Cool! I've been busy with work and school but should finish this by the end of next week.

**Comment by vuki656** [comment]
Is it expected for luacheck to lint markdown files?

**Comment by MunifTanjim** [comment]
> Is it expected for luacheck to lint markdown files? Not sure what the fix would be.

Don't worry about those. I've fixed the lint command on `main` branch. I'll rebase the PR before merging that would fix it.

I'll review the PR one last time as soon as possible... I'll probably make some minor changes before merging, if that's okay with you 😃 (or if you want to do those yourself I can add some more comments too)

**Comment by vuki656** [comment]
> > Is it expected for luacheck to lint markdown files? Not sure what the fix would be.
> 
> Don't worry about those. I've fixed the lint command on `main` branch. I'll rebase the PR before merging that would fix it.
> 
> I'll review the PR one last time as soon as possible... I'll probably make some minor changes before merging, if that's okay with you  (or if you want to do those yourself I can add some more comments too)

Oh, okay.

Whichever route would be easier for you. If you prefer just leaving comments ill take care of it, otherwise feel free to tweak whatever you'd like.

**Review by MunifTanjim — APPROVED** [review]

**Comment by MunifTanjim** [comment]
Thanks again for revamping the whole documentation 😃 🎉 🚀

**Pull request merged**
